### PR TITLE
[TS migration] Migrate 'withPolicy.js' HOC to TypeScript

### PR DESCRIPTION
**Pull request by blazejkustra**
### Details

Migration of 'withPolicy.js' to Typescript.

### Fixed Issues
$ https://github.com/Expensify/App/issues/24965
PROPOSAL: N/A

### Tests
- [x] Verify that no errors appear in the JS console
- [x] Verify that Workspace screens work as before (`withPolicy` is used in multiple components inside workspaces)

### Offline tests
N/A

### QA Steps
Same as Tests

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/46095609/03819036-b410-4fd9-b351-270d089f1199

</details>

<details>
<summary>Mobile Web - Chrome</summary>

[a-web.webm](https://github.com/Expensify/App/assets/46095609/0f45d419-d69c-42b1-a84a-f4f17dc202e8)

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/46095609/d1003779-369c-4483-b280-21b605666356

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/46095609/fb771baf-3e9d-4308-8ec5-bc6c41ee2c6b

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/46095609/0c5279a2-956e-4986-9eec-e356114f2962

</details>

<details>
<summary>Android</summary>

[android2.webm](https://github.com/Expensify/App/assets/46095609/8918e19c-a482-48ac-87a1-84226ffad07f)

</details>

### CODE REVIEW

File: src/pages/workspace/withPolicy.tsx [hunk 1]
```diff
@@ -61,59 +62,51 @@ const policyPropTypes = {
...
-        if (_.isString(policyID) && !_.isEmpty(policyID)) {
+export default function withPolicy<TProps extends WithPolicyProps, TRef>(
+    WrappedComponent: ComponentType<TProps & RefAttributes<TRef>>,
+): React.ComponentType<Omit<TProps, keyof WithPolicyOnyxProps>> {
+    function WithPolicy(props: TProps, ref: ForwardedRef<TRef>) {
+        const routes = useNavigationState((state) => state.routes || []);
+        const currentRoute = routes?.[routes.length - 1];
+        const policyID = getPolicyIDFromRoute(currentRoute as PolicyRoute);
+
+        if (typeof policyID === 'string' && policyID.length > 0) {
```
Comment (BartoszGrajdek) [inline src/pages/workspace/withPolicy.tsx:93]: This typecheck seems redundant as `getPolicyIDFromRoute` always returns string

File: src/pages/workspace/withPolicy.tsx [hunk 2]
```diff
@@ -1,21 +1,22 @@
...
+    params?: {
+        policyID: string;
+    };
+};
+
+function getPolicyIDFromRoute(route: PolicyRoute): string {
+    return route?.params?.policyID ?? '';
 }
 
 const policyPropTypes = {
```
Comment (BartoszGrajdek) [inline src/pages/workspace/withPolicy.tsx:22]: Since we're already leaving this for places where it is being used, maybe let's create an alternative written in TS for future use-cases once other files get migrated to Typescript?

File: src/pages/workspace/withPolicy.tsx [hunk 3]
```diff
@@ -61,59 +62,51 @@ const policyPropTypes = {
     policyMembers: PropTypes.objectOf(policyMemberPropType),
 };
 
-const policyDefaultProps = {
-    policy: {},
+const policyDefaultProps: WithPolicyOnyxProps = {
```
Comment (BartoszGrajdek) [inline src/pages/workspace/withPolicy.tsx:65]: NAB: `WithPolicyOnyxProps` is being used before it is being declared, maybe let's move it for the sake of clarity? :)

Comment (blazejkustra) [reply src/pages/workspace/withPolicy.tsx:65]: Changed!

**Review by BartoszGrajdek — APPROVED** [review]

**Review by BartoszGrajdek — APPROVED** [review]

File: src/pages/workspace/withPolicy.tsx [hunk 4]
```diff
@@ -61,59 +58,51 @@ const policyPropTypes = {
...
-        const currentRoute = _.last(useNavigationState((state) => state.routes || []));
-        const policyID = getPolicyIDFromRoute(currentRoute);
-
-        if (_.isString(policyID) && !_.isEmpty(policyID)) {
+export default function withPolicy<TProps extends WithPolicyProps, TRef>(
+    WrappedComponent: ComponentType<TProps & RefAttributes<TRef>>,
+): React.ComponentType<Omit<TProps, keyof WithPolicyOnyxProps>> {
+    function WithPolicy(props: TProps, ref: ForwardedRef<TRef>) {
+        const routes = useNavigationState((state) => state.routes || []);
+        const currentRoute = routes?.[routes.length - 1];
```
Comment (fabioh8010) [inline src/pages/workspace/withPolicy.tsx:87]:
```suggestion
        const currentRoute = routes?.at(-1);
```

File: src/pages/workspace/withPolicy.tsx [hunk 5]
```diff
@@ -61,59 +58,51 @@ const policyPropTypes = {
...
+                {...props}
+                ref={ref}
             />
         );
     }
 
-    WithPolicy.propTypes = propTypes;
-    WithPolicy.defaultProps = defaultProps;
     WithPolicy.displayName = `withPolicy(${getComponentDisplayName(WrappedComponent)})`;
-    const withPolicy = React.forwardRef((props, ref) => (
```
Comment (fabioh8010) [inline src/pages/workspace/withPolicy.tsx:108]: Why did you remove the `forwardRef`?

**Review by fabioh8010 — APPROVED** [review]
LGTM

**Comment by blazejkustra** [comment]
@Ollyws I resolved the conflicts. You can work on the reviewer checklist now, thanks! 😄

**Review by Ollyws — APPROVED** [review]
LGTM

**Review by joelbettner — APPROVED** [review]

File: src/pages/workspace/withPolicy.tsx [hunk 6]
```diff
@@ -61,61 +57,49 @@ const policyPropTypes = {
...
  * HOC for connecting a policy in Onyx corresponding to the policyID in route params
  */
-export default function (WrappedComponent) {
-    const propTypes = {
-        /** The HOC takes an optional ref as a prop and passes it as a ref to the wrapped component.
-         * That way, if a ref is passed to a component wrapped in the HOC, the ref is a reference to the wrapped component, not the HOC. */
-        forwardedRef: PropTypes.func,
+export default function <TProps extends WithPolicyProps, TRef>(WrappedComponent: ComponentType<TProps & RefAttributes<TRef>>): React.ComponentType<Omit<TProps, keyof WithPolicyOnyxProps>> {
+    function WithPolicy(props: TProps, ref: ForwardedRef<TRef>) {
+        const routes = useNavigationState((state) => state.routes || []);
```
Comment (roryabraham) [inline src/pages/workspace/withPolicy.tsx:83]: @blazejkustra why did we use `useNavigationState` here to get the current route instead of just using `useRoute`?

Comment (blazejkustra) [reply src/pages/workspace/withPolicy.tsx:83]: I reused the logic that was written before, with Typescript migration I usually try to avoid refactoring that could cause regressions, also at the time navigation wasn't migrated yet so I just didn't want to touch it here 😄 
```ts
const currentRoute = _.last(useNavigationState((state) => state.routes || []));
```
